Closure elimination for new backend

lrytz opened the pull request:
When an indylambda closure is allocated and invoked within the same method, rewrite the invocation to the implementation method.

lrytz committed:
Producers / Consumers Analysis

ASM has a built-in `SourceValue` analysis which computes for each
value a set of instructions that may possibly have constructed it.

The ProdConsAnalyzer class provides additional queries over the
result of the SourceValue analysis:
  - consumers of values
  - tracking producers / consumers through copying operations (load,
    store, etc)

A fix to (and therefore a new version of) ASM was required. See here:
https://github.com/scala/scala-asm/commit/94106a5472

lrytz commented:
Ready for a first round of feedback. @retronym, @SethTisue

retronym commented:
I gave this an in-person review on Tuesday. I thought @lrytz added most of my comments to as GitHub comments, but I don't see them here. AFAICT, the first commit has incorporated my comments, but the second one still needs to be polished.

The tests for prod/cons analysis are nicely structured.

The IDE build failure is unrelated to this patch:

```
[sbt-republish] [info] Updating {file:/home/jenkins/workspace/scala-2.11.x-integrate-ide/target/zinc/target-0.9.1/project-builds/sbt-republish-40094874e9a4cc023749c371f693ef67a5da2d41/project/project/}sbt-republish-40094874e9a4cc023749c371f693ef67a5da2d41-build-build...
[sbt-republish] [warn]  [FAILED     ] org.apache.maven#maven-compat;3.0.4!maven-compat.jar: invalid sha1: expected=14d626e1e29a36f5f4629b689e964f8665707839 computed=211dc5b40b735d75fd7ac245a0ca5e82a3ab1508 (3637ms)
[sbt-republish] [warn]  [FAILED     ] org.apache.maven#maven-compat;3.0.4!maven-compat.jar: invalid sha1: expected=14d626e1e29a36f5f4629b689e964f8665707839 computed=211dc5b40b735d75fd7ac245a0ca5e82a3ab1508 (3637ms)
[sbt-republish] [warn] ==== public: tried
[sbt-republish] [warn]   http://repo1.maven.org/maven2/org/apache/maven/maven-compat/3.0.4/maven-compat-3.0.4.jar
[sbt-republish] [warn]  [FAILED     ] org.jsoup#jsoup;1.6.1!jsoup.jar: invalid sha1: expected=70f0dfbe511d84ee7d0dda238a52746751787c47 computed=f124031a316ad95c6498ca5d97e7bcf0fffa00d1 (1941ms)
[sbt-republish] [warn]  [FAILED     ] org.jsoup#jsoup;1.6.1!jsoup.jar: invalid sha1: expected=70f0dfbe511d84ee7d0dda238a52746751787c47 computed=f124031a316ad95c6498ca5d97e7bcf0fffa00d1 (1941ms)
```

lrytz commented:
@retronym indeed, i cleaned up the first commit, but still working on the second. should appear soon. my self-comments are in this revision: https://github.com/lrytz/scala/commit/0b7e0f108bc2406933c45e2fd70912543410c34b

lrytz committed:
Rewrite closure invocations to the lambda body method

When an indylambda closure is allocated and invoked within the same
method, rewrite the invocation to the implementation method.

This works for any indylambda / SAM type, not only Scala functions.
However, the Scala compiler (under -Xexperimental) currently desugars
function literals for non-FunctionN types to an anonymous class during
typer.

No testing yet, waiting for FunctionN to become SAMs first.

The feature requires scala-java8-compat to be on the classpath and a
number of compiler flags:
-Ydelambdafy:method -Ybackend:GenBCode -Yopt:closure-elimination -target:jvm-1.8

➜  scala git:(opt/closureInlining) ant -Dscala-java8-compat.package=1 -Dlocker.skip=1
➜  scala git:(opt/closureInlining) cd sandbox

➜  sandbox git:(opt/closureInlining) cat Fun.java
public interface Fun<T> {
  T apply(T x);
}
➜  sandbox git:(opt/closureInlining) javac Fun.java

➜  sandbox git:(opt/closureInlining) cat Test.scala
class C {
  val z = "too"
  def f = {
    val kap = "me! me!"
    val f: Tuple2[String, String] => String = (o => z + kap + o.toString)
    f(("a", "b"))
  }

  def g = {
    val f: Int => String = x => x.toString
    f(10)
  }

  def h = {
    val f: Fun[Int] = x => x + 100 // Java SAM, requires -Xexperimental, will create an anonymous class in typer
    f(10)
  }

  def i = {
    val l = 10l
    val f: (Long, String) => String = (x, s) => s + l + z + x
    f(20l, "n")
  }

  def j = {
    val f: Int => Int = x => x + 101 // specialized
    f(33)
  }
}
➜  sandbox git:(opt/closureInlining) ../build/quick/bin/scalac -target:jvm-1.8 -Yopt:closure-elimination -Ydelambdafy:method -Ybackend:GenBCode -Xexperimental -cp ../build/quick/scala-java8-compat:. Test.scala

➜  sandbox git:(opt/closureInlining) asm -a C.class
➜  sandbox git:(opt/closureInlining) cat C.asm

[...]

  public g()Ljava/lang/String;
   L0
    INVOKEDYNAMIC apply()Lscala/compat/java8/JFunction1; [
      // handle kind 0x6 : INVOKESTATIC
      java/lang/invoke/LambdaMetafactory.altMetafactory(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;[Ljava/lang/Object;)Ljava/lang/invoke/CallSite;
      // arguments:
      (Ljava/lang/Object;)Ljava/lang/Object;,
      // handle kind 0x6 : INVOKESTATIC
      C.C$$$anonfun$2$adapted(Ljava/lang/Object;)Ljava/lang/String;,
      (Ljava/lang/Object;)Ljava/lang/String;,
      3,
      1,
      Lscala/Serializable;.class,
      0
    ]
    CHECKCAST scala/Function1
   L1
    ASTORE 1
   L2
    ALOAD 1
    BIPUSH 10
    INVOKESTATIC scala/runtime/BoxesRunTime.boxToInteger (I)Ljava/lang/Integer;
    ASTORE 2
    POP
    ALOAD 2
    INVOKESTATIC C.C$$$anonfun$2$adapted (Ljava/lang/Object;)Ljava/lang/String;
    CHECKCAST java/lang/String
   L3
    ARETURN

[...]

lrytz commented:
@retronym done with the cleanup. Summary of changes since last review:

ProdCons
- ProdCons correctly handles instructions that produce multiple values (e.g., `DUP`).
- ProdCons supports all exotic copying instructions (e.g., `DUP2_X2`) for queries that thread through copying instructions (`initialProducers..` / `ultimateConsumers..`)
- fixed ProdCons for `INSTANCEOF`
- use `AnyRefMap` for mutable maps
- more testing

Closure apply rewriting
- created `Handle` objects to compare against for `metafactory` / `altMetafactory` (instead of comparing all fields). Similar for the implMethod, created a Type to compare against. See `isClosureInstantiation`.
- cleaned up the trickery with the opCode-offset, better comments
- run ProdCons only if it's actually needed, i.e., if there's some callsite within the method that looks like a closure apply call (in terms of name and descriptor).
- support all calling conventions (`NEWINVOKESPECIAL` etc)

retronym commented on src/compiler/scala/tools/nsc/backend/jvm/BackendReporting.scala:
```diff
@@ -247,6 +247,28 @@ object BackendReporting {
+   * Used in `rewriteClosureApplyInvocations` when a closure apply callsite cannot be rewritten
+   * to the closure body method.
+   */
+  trait RewriteClosureApplyToClosureBodyFailed extends OptimizerWarning {
```
sealed?

retronym commented:
Regarding the using or pre-allocated tuples in `InstructionStackEffect` (`def t`), I think that `IntMap#contains` actually allocates an `Option` as it is implemented in terms of `get(k).isDefined`.

retronym commented:
LGTM as a starting point. We'll need to add automated tests once we have the indylambda functional interfaces in the repo, which should do this week, using your patch to embed the `JFunctionN` sources.

Have you performed any manual testing with indylambda and closure inlining enabled for a big codebase (e.g. boostrapping scala)?

lrytz commented:
good piont about `IntMap`.. Running bootstrap testing now.

lrytz commented:
found one issue:

```
    val fun: String => String = l => l
    val l = List("")
    fun(l.head)
```

Here, `l.head` has type `Object`, but the lambda body function expects a `String`, so I get

```
java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    Test$.f()Ljava/lang/String; @42: invokestatic
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to 'java/lang/String'
```

retronym commented:
I guess we'll need the type adaptation framework after all.

lrytz commented:
i think we can just cast in these cases

lrytz committed:
Cast arguments where necessary before rewriting closure invocations

The parameter types of a closure invocation can be supertypes of the
parameter types in the implementation method. The closure automatically
casts arguments to the right type. We need to do the same when
rewriting closure invocations. Example:

    val fun: String => String = l => l
    val l = List("")
    fun(l.head)

The closure object calls `apply(Object)Object`. The body method takes
an argument of type `String`.

lrytz commented:
Added another commit. The test suite seems to run fine now (some unrelated / expected failures due to indyLambda, closure rewriting). Also `strap.done` / `test.stability` work.

retronym commented:
LGTM! Great work!
